moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp

```diff
@@ -391,53 +403,91 @@ void planning_scene_monitor::CurrentStateMonitor::jointStateCallback(const senso
           robot_state_.setJointPositions(jm, &b.max_position_);
       }
     }
+  }
 
-    // read root transform, if needed
-    if (tf_ && (robot_model_->getRootJoint()->getType() == robot_model::JointModel::PLANAR ||
-                robot_model_->getRootJoint()->getType() == robot_model::JointModel::FLOATING))
+  // callbacks, if needed
+  if (update)
+    for (std::size_t i = 0; i < update_callbacks_.size(); ++i)
+      update_callbacks_[i](joint_state);
+
+  // notify waitForCurrentState() *after* potential update callbacks
+  state_update_condition_.notify_all();
+}
+
+void planning_scene_monitor::CurrentStateMonitor::tfCallback()
+{
+  bool update = false;
```

Comment:
I don't see a difference between `update` and `changes`.

Reply:
The implementation of the monitor always triggers the update signal at the end of the JointState callback, because some thread might be waiting for the joint states at the current time.
But it calls the `update_callbacks_` only if something actually *changed* (see [here](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp#L357)).

With the tf callback we can't even be sure the time of any joint updated and we don't want to trigger the notify unless we actually updated *something*.
This is represented in `bool update`.

`bool changes` represents whether something actually changed.
Because Multi-DOF joints tend to be rather noisy, I decided to use a threshold instead of just comparing the values.
Afair the update callbacks in MoveIt mark the planning scene as changed which can trigger safety stops in currently executing plans, re-validation of current plans, and publishing of the monitored_planning_scene topic. So unnecessary update callbacks can cause a lot of overhead.

Reply:
Sorry, still don't get the difference: I guess you update something only if it has changed, don't you?

Reply:
Got it now from reading the code: If there is only a tiny change, you apply this change, but you don't want to call the callbacks.